RFC [Reject]: Boot Orchestration Component

kat-perez opened the pull request:
## Description

This RFC proposes modular components for UEFI spec compliant boot functionality in Patina firmware. Platforms can
use reference implementations for console initialization, boot option discovery, and boot orchestration,
or implement custom boot flows using the provided library functions.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [x] Includes documentation?
- [ ] Creates a new crate?

## How This Was Tested

N/A

## Integration Instructions

N/A

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+1. **Functional Equivalence**: The Rust implementation provides all functionality required by the UEFI Boot Manager
+2. **Semantic Preservation**: Boot sequence, variable handling, and platform callbacks maintain identical behavior
+3. **UEFI Compliance**: Full adherence to UEFI 2.11 and PI 1.9 specifications (December 2024)
+4. **Platform Compatibility**: Supports all 15+ EDK2 platforms currently using BdsDxe with optional feature customization
```
Many many more platforms use EDK II than 15 (that may be the number in edk2-platforms). This line probably can be changed to something like "Supports platform extensibility to meet current scenarios"

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
We will want to review this BootManagerConfig very closely. We will not want to match all of EDK II's configuration options. We should have a very good reason for each configuration option we come up with.

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
Similarly, we should be crystal clear on what platform hooks we are adding and why.

kat-perez replied:
I want to propose that platform does their own platform customization for BootOrchestrator component instead of doing exactly what BdsDxe does with platform callbacks

kuqin12 commented on docs/src/rfc/text/0000-boot-manager.md:
This does not seem an optional service.

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
Is the intention that the platform must provide the PlatformPolicy service? So that the component depends on that being available and as such just calls to it?

kat-perez replied:
i've removed this mechanism since I went a different direction with BootOrchestration

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
This seems like it could be decoupled from BootManager?

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
I would weigh whether this is better as a trait or as a service. It depends on how much functionality lives behind this. It may be a trait is better here because there will be no other consumers of it. I need to think through the scenario a bit more.

kat-perez replied:
I think I can do more of an analysis of this in the "Unresolved Questions" section and propose some ideas there

os-d commented on docs/src/rfc/text/0000-boot-manager.md:
I may be confused here, are you saying no option ROM dispatch at all? Or a certain scenario that is removed? We definitely have to support option ROMs, despite the problems they come with.

kat-perez replied:
I left option ROM dispatch out of the scope of this because it occurs during PCI device enumeration in the DXE phase, before Boot Manager execution. I think it's not a required feature for an ideal Boot Manager. I think we would need an OptionRomDispatcher component instead which is outside of the scope of this RFC.

os-d reviewed (commented):
The biggest part of BDS that I did not see enumerated in this RFC is the connect management, driving binding protocols, etc. That requires careful thought as it is another form of driver dispatch that is unique to the BDS phase.

I think that before we move this forward, we need to include what that story is.

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
The RFC makes a lot of references to `BdsDxe`. I think that should be mostly consolidated to a single section focused on that relationship so that the remainder of the document is free to describe a new design without that coupling.

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+# RFC: Boot Manager
+
+This RFC proposes a modular Boot Manager for Patina as a drop-in replacement for EDK2's BdsDxe driver. The design
+transforms the monolithic 1200-line EDK2 BdsDxe into a service-oriented architecture with 3 required core services
```
The line count of `BdsDxe` itself is not that relevant here. Its practical complexity is significantly extended with iibraries like `PlatformBootManagerLib` and `UefiBootManagerLib`.

kat-perez replied:
removed any language like this and comparisons to EDK II BdsDxe

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+5. **Platform Customization Complexity**: Library override patterns require rebuilding entire firmware stacks
+6. **Binary Bloat**: All features compiled in whether needed or not
+
+**Compatibility Definition** (consistent with other Patina RFCs):
```
What does "consistent with other Patina RFCs" mean?

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+- **UefiBootManagerLib** - Boot option management library
+- **PlatformBootManagerLib** - Platform customization callbacks
+- **EFI_BDS_ARCH_PROTOCOL** - BDS architectural protocol
+- **Variable Policy Protocol** interactions
```
My understanding is that this is saying this implemenation will interact with an installed variable policy protocol. If that's the case, the "protocol" detail is not that important and I'd update it to just say it will properly set variable policies.

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+## Requirements
+
+1. **Drop-in Replacement**: Enable platforms to remove BdsDxe.inf from their build and use Patina Boot Manager
+   without modifying other firmware components
```
Yeah, I don't think it is necessary to constrain the design to be a drop-in replacement for a specific C implementation. Instead, the compability point should be the UEFI Specification.

kuqin12 commented on docs/src/rfc/text/0000-boot-manager.md:
I was going to suggest on the protocol dependencies, but my guess is that it will be just those arch protocols, we probably cannot do much better than that.

However, this `PlatformBootManager` abstract might contain a lot of content, and some are very much in common. i.e. the signal end of dxe, dispatch deferred images. Would consider redoing the abstraction with this RFC?

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
Some of these items like "memory safety" I don't think are requirements or at least as written. For example, it is implied that if the code is written in Rust these memory safety guarantees will be present.

If there are specific statements being made about something in particular like certain C FFI interfaces or something those should be called out specifically.

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
I don't think a "BdsDxe" analysis section is needed. This should focus on a design and only reference minor implementation points necessary in the BdsDxe implementation.

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
We generally do not want to tangle an EDK II invention (PCDs) with Patina. Please clearly call out the separation of configuration for Pure Rust builds.

kat-perez replied:
removed PCDs from this RFC

makubacki commented on docs/src/rfc/text/0000-boot-manager.md:
After having read the RFC, at a high-level these are some broader changes that I think can bring focus.

1. The goal of the RFC is torn throughout by (1) trying to support Rust features and the Patina component model and (2) being a "drop-in replace for `BdsDxe`" and bringing in C-based and EDK constructs. Clearly scoping this is important and should be addressed directly. Please clearly articulate why the scope of the RFC encompasses both of these. If the goal is to support both, you have three pieces to the design that occur - a "common design", "edk2-specific considerations like PCDs and protocols", and "Rust/Patina-specific considerations like services". At nearly every point in the RFC, clarity is needed on how these different use cases are supported by what's mentioned. In some cases, it might be obvious, there might be overlap, and in other cases not.
2. Partially due to the first point, I think the deep tethering to `BdsDxe` throughout the entire document encumbers a new design based on Patina components and services. Though, the suggestion for clarification in the first point might help this. If the goal is to support an independent "drop-in for BdsDxe" (implying the rest of Patina is not present) and a component for Patina, the RFC is going to get complex.
3. A lot of information is repeated throughout the document. Please identify and remove redundant points.

I think this document should primarily focus on what an ideal UEFI "boot manager" design looks like, why, and how it is best accomplished using Rust features and the Patina integration model.

kat-perez replied:
Agree that I think I should rewrite the entire RFC based on the direction towards an ideal UEFI boot manager design. Will remove the goal of being a "drop-in" replacement for BdsDxe and focus more on the Patina integration model.

kat-perez commented:
i've changed the design quite a bit, added more goals and requirements to further constrain the design to be limited to the most fundamental Patina components that can achieve a successful boot

kat-perez commented on docs/src/rfc/text/0000-boot-manager.md:
```diff
@@ -0,0 +1,707 @@
+callbacks, hotkeys, OS indications, and boot execution—not just device selection. The "BDS" name is a historical EDK2
+artifact that doesn't reflect the full scope of functionality.
+
+Additional scope includes:
```
update: removed these services for now for the scope of this RFC but add them to the optional features so that a separate RFC be required for justifying why some of these features would need to be components/services

kat-perez commented on docs/src/rfc/text/0000-boot-manager.md:
removed this question and then also updated the unresolved questions

kat-perez commented on docs/src/rfc/text/0000-boot-manager.md:
removed PCD reads

apop5 commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,551 @@
+This RFC provides three reference components for console initialization, boot option discovery, and boot
+orchestration. Platforms compose these components to create standard or custom boot flows. Library
+functions support custom implementations without code duplication.
+
```
We are missing information about events that are signaled during BDS.

Bds is responsible for the End of Dxe event, and a lot of drivers rely on this signal (https://github.com/tianocore/edk2/blob/master/StandaloneMmPkg/Core/StandaloneMmCore.c#L71)

There are other events that are typically signaled during the bds phase, and current EDK2 relies on providing plaform level drivers that link into bds to signal these events.

If we are going to be able to provide a compatible rust BDS, I think these signaled events need to be covered in the RFC so that compatability can be determine by users as they look to switch from edk2

joschock commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+
+## StandardBootManager Component
+
+`StandardBootManager` is a composite component that provides complete UEFI-compliant boot functionality
```
super-nit: naming: since this is a minimal config simple platform case, consider `SimpleBootManager` instead of `Standard` - which IMHO implies "does most of the things most standard platforms want" which I don't think is in view here.

kat-perez replied:
I've changed two things I'm using `BootOrchestrator` to be the full standard implementation and changed this to `SimpleBootOrchestrator` to refer to a minimal platform solution (also serves as a sample of a custom `BootOrchestrator`)

Reframed:
  - Before: "reference implementation... suitable for development and simple platforms. Production platforms typically implement custom"
  - After: "standard implementation... suitable for most platforms. Platforms with specific requirements can implement custom"

joschock commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+
+#### Component Decomposition
+
+The three components map to distinct phases of UEFI boot. **Console initialization** populates console
```
Maybe some commentary on why this three-way decomposition is chosen. Other decompositions could be imagined with more steps (for example, you could break out **System Bus Initialization** from **Console initialization** and stick PCI enumeration in there). I'm not saying the three-way is wrong, necessarily, but would be useful to have a rationale for why that particular decomposition is preferred.

joschock commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+
+### Connection Strategy Pattern
+
+BootOptionDiscovery accepts a connection strategy as configuration to control which BlockIo device
```
`BlockIo` is a bit limiting. I think ultimately, you might want to look at anything with a filesystem you support regardless of the underlying storage foundation. Such a files system might be exposed via means other than an underlying `BlockIo` (though that is the most traditional way for fixed boot media).

kat-perez replied:
Agreed on this. I've updated the RFC so that it no longer just specifies BlockIo or any particular storage foundation. This work is deferred to BootOptionDiscovery component

os-d commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+## Goals
+
+1. Provide UEFI spec compliant boot functionality
+2. Enable platforms to boot with zero custom code using reference components
```
This statement may read as a platform does not need any custom code to boot, i.e. just taken patina as is and you are good, which is not what you intend. Consider rewriting this as something like:

`Enable platforms to use a common boot manager without customization`

kat-perez replied:
Goal #2 reworded like this:
> 2. Enable platforms to use a common boot manager without customization

kat-perez commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+
+To discover bootable media, the Boot Manager connects storage device controllers (via `ConnectController()`) to
+enable file system drivers and scan for boot loaders. BootOptionDiscovery accepts connection strategies as
+configuration, allowing platforms to customize which devices get connected while using the shared ProtocolServices
```
good point. I've added an explicit "Future Work: Connect Controller Service" section noting that this is deferred and native Patina components cannot participate in the connect controller flow in this RFC.

kat-perez commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,640 @@
+**ConsoleDiscovery**: Scans for GOP and SimpleTextInput protocol handles, creates device paths, and
+writes `ConIn`/`ConOut`/`ErrOut` variables. Depends on RuntimeVariableServices and ProtocolServices.
+
+**BootOptionDiscovery**: Discovers bootable media by connecting storage controllers and scanning file
```
I think the "BootOptionDiscovery" component can be inclusive of all possible types of "boot options".  I'm leaving the role of Config<BootOptions> to take device paths and just bypass the whole Boot#### / BootOrder variables.  That supports a hardcoded device path directly in config

kat-perez commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,510 @@
+
+## Unresolved Questions
+
+### BDS Phase Dispatch Mechanism
```
@os-d @makubacki @joschock @Javagedes, added this new question in "Unresolved Questions" that I would like feedback on please

joschock replied:
Yep - that looks reasonable. My main concern is that patina doesn't become tightly coupled with the boot-orchestrator. building one that boots legacy BDS should keep us honest :)

os-d commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,510 @@
+event signaling.
+
+For platforms transitioning from C-based BDS, `call_bds()` could be retained as a fallback that
+only executes if no Patina boot component ran.
```
I think that another way to approach this is to have an extremely minimal BDS component that just calls the BDS arch protocol. So, if a platform starts with not using the `BootOrchestrator` component, it can use the `DummyBdsComponent` or whatever which will just call the protocol to engage `BdsDxe`. Then Patina can be agnostic of what owns BDS, it simply has one way of initiating the handoff (kind of) to BDS.

rogurr commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,735 @@
+- **`Key####` hotkey variables**: Full UEFI-compliant hotkey management via `Key####`
+  variables, extending the current scancode-based `detect_hotkey()` to read hotkey
+  configuration from UEFI variables.
+- **Platform recovery options**: `PlatformRecovery####` variable support for
```
Could this be expanded to maybe be a hierarchy mechanism with a fallover?  Try option 1, if fails, try option 2, if fails, try option 3, etc.?  We talked about the recovery option which would be where we just have 2 options, but current platforms have several options; Windows boot path, Linux boot path, UEFI shell, PXE boot, USB boot, front page, etc.

kat-perez replied:
Yeah, that makes sense. I've updated it to describe a fallback chain supporting arbitrary depth

rogurr commented on docs/src/rfc/text/0000-patina-boot.md:
```diff
@@ -0,0 +1,735 @@
+
+Both `expand_device_path()` and `boot_from_device_path()` are provided to support different use cases:
+
+| Function                   | Use Case                                                                                                                                 |
```
Minor, if you want to make tables a little more readable in edit mode, you don't need to make each line the same length.

This text:
``` text
| Title 1 Left Justified | Title 2 Centered | Title 3 Right Justified |
|---|:---:|---:|
| a | b | c |
| d | e | f |
```

Produces this:
| Title 1 Left Justified | Title 2 Centered | Title 3 Right Justified |
|---|:---:|---:|
| a | b | c |
| d | e | f |

The separator only needs 3 dashes and you can put a colon ':' on both sides to center the column or on the right to right align.

kat-perez committed:
RFC [Reject]: Boot Orchestration Component

makubacki reviewed (approved):
Thank you for all of your work on this design and implementation. Approving the update to move it to rejected status.
